plutus-use-cases/scripts/Main.hs

```diff
@@ -0,0 +1,82 @@
+import           System.Environment                                    (getArgs)
+import           System.FilePath                                       ((</>))
+import qualified Wallet.Emulator.Folds                                 as Folds
+import           Wallet.Emulator.Stream                                (defaultEmulatorConfig, foldEmulatorStreamM)
+import qualified Wallet.Emulator.Stream                                as Stream
+
+import qualified Language.PlutusTx.Coordination.Contracts.Crowdfunding as Crowdfunding
+import qualified Language.PlutusTx.Coordination.Contracts.Currency     as Cur
+import qualified Language.PlutusTx.Coordination.Contracts.Game         as Game
+import qualified Spec.Currency                                         as Currency
+import qualified Spec.Escrow                                           as Escrow
+import qualified Spec.Future                                           as Future
+import qualified Spec.GameStateMachine                                 as GameStateMachine
+import qualified Spec.MultiSig                                         as MultiSig
+import qualified Spec.MultiSigStateMachine                             as MultiSigStateMachine
+import qualified Spec.PingPong                                         as PingPong
+import qualified Spec.Prism                                            as Prism
+import qualified Spec.PubKey                                           as PubKey
+import qualified Spec.Stablecoin                                       as Stablecoin
+import qualified Spec.TokenAccount                                     as TokenAccount
+import qualified Spec.Vesting                                          as Vesting
+
+main :: IO ()
+main = do
+    args <- getArgs
+    case args of
+        [script_path] -> writeScripts script_path
+        _             -> putStrLn "usage: plutus-use-cases-scripts SCRIPT_PATH"
+
+writeScripts :: FilePath -> IO ()
+writeScripts fp = do
+    putStrLn $ "Writing scripts to: " <> fp
+    traverse_ (uncurry (writeScriptsTo fp))
+        [ ("crowdfunding-success", Crowdfunding.successfulCampaign)
+        , ("currency", Currency.currencyTrace)
+        , ("escrow-redeem_1", Escrow.redeemTrace)
+        , ("escrow-redeem_2", Escrow.redeem2Trace)
+        , ("escrow-refund", Escrow.refundTrace)
+        , ("future-increase-margin", Future.increaseMarginTrace)
+        , ("future-settle-early", Future.settleEarlyTrace)
+        , ("future-pay-out", Future.payOutTrace)
+        , ("game-guess", Game.guessTrace)
+        , ("game-guessWrong", Game.guessWrongTrace)
+        , ("game-sm-success", GameStateMachine.successTrace)
+        , ("game-sm-success_2", GameStateMachine.successTrace2)
+        , ("multisig-success", MultiSig.succeedingTrace)
+        , ("multisig-failure", MultiSig.failingTrace)
+        , ("multisig-sm_1", MultiSigStateMachine.lockProposeSignPay 3 1)
+        , ("multisig-sm_2", MultiSigStateMachine.lockProposeSignPay 3 3)
```

Comment:
Just a remark, but I found a lot more distinct validations when I was experimenting with the use cases: for example in the multisig state machine example ten different combinations of scripts appeared (see https://github.com/input-output-hk/plutus/tree/master/plutus-benchmark/validation/data/multisigSM).  I just got those by writing out all the scripts appearing in `Ledger.Scripts.runScript` and manually eliminating duplicates.  Admittedly I was never able to work out exactly which stages of the contract those corresponded to!  Maybe some of them were for subsidiary transactions?

Reply:
Thanks for the link, I didn't know about that README.

`lockProposeSignPay k n` should result in up to _(k+2)*n_ script validations - there are _n_ rounds, and for each round we have _k_ steps where a signature is added, 1 step for the proposal and 1 step for the payment transition. The actual number may be smaller because validation stops after the first failing script is encountered.

Actually I think that `lockProposeSignPay 3 2` covers all ten variations you got.

The cases where the transition is invalid (for example `lockProposeSignPay 2 1`) don't result in an invalid transaction because the client realises that the transaction would fail and doesn't submit it.